components/storage/blockdevice/COMPONENT_SDIO/SDIOBlockDevice.cpp

```diff
@@ -0,0 +1,398 @@
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_PARAMETER;
+    }
+
+    // HAL layer uses uint32_t for addr/size
+    uint32_t *_buffer = (uint32_t *)(buffer);
+
+    // Get block count
+    bd_size_t blockCnt = size / _block_size;
+    addr = addr / _block_size;
+
+#if DEVICE_SDIO_ASYNC
+    // make sure card is ready
+    {
+        uint32_t tickstart = us_ticker_read();
+        while (sdio_get_card_state() != SD_TRANSFER_OK) {
+            // wait until SD ready
+            if ((us_ticker_read() - tickstart) >= MBED_CONF_SDIO_CMD_TIMEOUT) {
+                unlock();
+                return SD_BLOCK_DEVICE_ERROR_WRITEBLOCKS;
+            }
+        }
+    }
+
+    status = sdio_writeblocks_async(_buffer, addr, blockCnt);
+    debug_if(SD_DBG, "WriteBlocks dbgtest addr: %lld  blockCnt: %lld \n", addr, blockCnt);
+
+    if (status == MSD_OK) {
+        // wait until DMA finished
+        uint32_t tickstart = us_ticker_read();
+        while (sdio_write_pending() != SD_TRANSFER_OK) {
+            if ((us_ticker_read() - tickstart) >= MBED_CONF_SDIO_CMD_TIMEOUT) {
+                unlock();
+                return SD_BLOCK_DEVICE_ERROR_WRITEBLOCKS;
+            }
+        }
+        // make sure card is ready
+        tickstart = us_ticker_read();
+        while (sdio_get_card_state() != SD_TRANSFER_OK) {
+            // wait until SD ready
+            if ((us_ticker_read() - tickstart) >= MBED_CONF_SDIO_CMD_TIMEOUT) {
+                unlock();
+                return SD_BLOCK_DEVICE_ERROR_WRITEBLOCKS;
+            }
+        }
+    } else {
+        debug_if(SD_DBG, "WriteBlocks failed! addr: %lld  blockCnt: %lld \n", addr, blockCnt);
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_WRITEBLOCKS;
+    }
+#else
+    status = sdio_writeblocks(_buffer, addr, blockCnt);
+
+    debug_if(SD_DBG, "WriteBlocks dbgtest addr: %lld  blockCnt: %lld \n", addr, blockCnt);
+
+    if (status != MSD_OK) {
+        debug_if(SD_DBG, "WriteBlocks failed! addr: %lld  blockCnt: %lld \n", addr, blockCnt);
+        status = SD_BLOCK_DEVICE_ERROR_WRITEBLOCKS;
+    }
+#endif
+
+    unlock();
+    return status;
+}
+
+int SDIOBlockDevice::trim(bd_addr_t addr, bd_size_t size)
+{
+    debug_if(SD_DBG, "trim Card...\r\n");
+    lock();
+    if (isPresent() == false) {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_DEVICE;
+    }
+
+    if (!_is_initialized) {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_INIT;
+    }
+
+    if (!_is_valid_trim(addr, size)) {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_PARAMETER;
+    }
+
+    bd_size_t blockCnt = size / _block_size;
+    addr = addr / _block_size;
+
+    int status = sdio_erase(addr, blockCnt);
+    if (status != 0) {
+        debug_if(SD_DBG, "Erase blocks failed! addr: %lld  blockCnt: %lld \n", addr, blockCnt);
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_ERASEBLOCKS;
+#if DEVICE_SDIO_ASYNC
+    } else {
+        uint32_t tickstart = us_ticker_read();
```

Comment:
why is this using _us_ticker - not wait_us() rather or  Ticker object? Shouldn't it `sleep_for` rather?